File: client/src-tauri/src/model/issue.rs
use crate::db::table::HasKey;
use crate::model::default_value::DefaultValue;
use crate::model::macro_model::model_with_default;
use crate::model::tag::Tag;
use crate::model::time_stamps::Timestamps;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[serde(rename_all = "PascalCase")]
pub enum IssueStatus {
    Open,       // 登録しただけ
    InProgress, // 対応中
    Review,     // レビュー待ち
    Resolved,   // 対応完了（レビュー済み）
    Closed,     // 完全クローズ
}

impl DefaultValue for IssueStatus {
    fn default_value() -> Self {
        IssueStatus::Open
    }
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[serde(rename_all = "PascalCase")]
pub enum IssuePriority {
    Low,
    Medium,
    High,
    Critical,
}

impl DefaultValue for IssuePriority {
    fn default_value() -> Self {
        IssuePriority::Low
    }
}

impl HasKey<(String, String)> for Issue {
    fn key(&self) -> (String, String) {
        (self.project_id.clone(), self.id.clone())
    }
}

model_with_default!(
    Issue,
    IssueRequest,
    {
        id: String,
        project_id: String,
    },
    {
        task_id: Option<String>,
        title: String,
        description: String,
        status: IssueStatus,
        priority: IssuePriority,
        owner: String,
        reviewer: String,
        due_date: Option<String>,
        completed_date: Option<String>,
        tags: Vec<Tag>,
    }
);

File: client/src/types/db/issue.ts
import { Tag, Timestamps } from "./common";

export enum IssueStatus {
    Open = "Open",
    InProgress = "InProgress",
    Review = "Review",
    Resolved = "Resolved",
    Closed = "Closed",
}

export enum IssuePriority {
    Low = "Low",
    Medium = "Medium",
    High = "High",
    Critical = "Critical",
}

export type IssueKey = {
    id: string;
    project_id: string;
};

export type IssueValue = {
    task_id?: string,
    title: string,
    description: string,
    status: IssueStatus,
    priority: IssuePriority,
    owner: string,
    reviewer: string,
    due_date?: string,
    completed_date?: string,
    tags: Tag[],
    timestamps: Timestamps,
};

// Row = Key + Value
export type IssueRow = IssueKey & IssueValue;
export type IssuePayload = IssueKey & Partial<IssueValue>;

export const toIssuePayload = (param: IssueRow): IssuePayload => (
    {
        id: param.id,
        project_id: param.project_id,
        task_id: param.task_id,
        title: param.title,
        description: param.description,
        status: param.status,
        priority: param.priority,
        owner: param.owner,
        reviewer: param.reviewer,
        due_date: param.due_date,
        completed_date: param.due_date,
        tags: param.tags,
    }
)
